Src/library/net/Clib/network.c

```diff
@@ -1225,37 +1192,47 @@ EIF_INTEGER c_get_sock_send_timeout(EIF_INTEGER fd, EIF_INTEGER level)
 #ifdef EIF_WINDOWS
 	int arg;
 	asize = sizeof(arg);
-	eif_net_check(getsockopt((EIF_SOCKET_TYPE) fd, (int) level, SO_SNDTIMEO, (char *) &arg, &asize));
-	return (EIF_INTEGER) (arg / 1000); /* Convert milliseconds to seconds */
+	eif_net_check(getsockopt((EIF_SOCKET_TYPE) fd, (int) level, optname, (char *) &arg, &asize));
+	return RTU64C(1000000) * arg; /* 1 millisecond = 1000000 nanoseconds */
 #else
 	struct timeval tv;
 	asize = sizeof(struct timeval);
-	eif_net_check(getsockopt((EIF_SOCKET_TYPE) fd, (int) level, SO_SNDTIMEO, &tv, &asize));
-	return (EIF_INTEGER) tv.tv_sec;
+	eif_net_check(getsockopt((EIF_SOCKET_TYPE) fd, (int) level, optname, &tv, &asize));
+	/* 1 second = 1000000000 nanoseconds, and 1 microsecond = 1000 nanoseconds */
+	return RTU64C(1000000000) * tv.tv_sec + RTU64C(1000) * tv.tv_usec;
 #endif
 }
 
-void c_set_sock_send_timeout(EIF_INTEGER fd, EIF_INTEGER level, EIF_INTEGER send_timeout_seconds)
-	/* set socket SO_SNDTIMEO option to send_timeout_seconds */
+void c_set_sock_opt_timeo(EIF_INTEGER fd, EIF_INTEGER level, int optname, EIF_NATURAL_64 timeout_nano_seconds)
+	/* set socket `optname` option to timeout_nano_seconds */
 {
 #ifdef EIF_WINDOWS
-	int arg = (int) 1000 * send_timeout_seconds; /* Timeout in milliseconds */
-	eif_net_check (setsockopt((EIF_SOCKET_TYPE) fd, (int) level, (int) SO_SNDTIMEO, (char *) &arg, sizeof(arg)));
+	int arg = (int)(timeout_nano_seconds / 1000000); /* Timeout in milliseconds: 1 millisecond = 1000000 nanoseconds */
```

Comment:
Do we care if there resulting value does not fit into an int?

Reply:
done